### ISSUE TITLE
Help with fine-tuning/transfer learning

### ISSUE THREAD

**Issue by nemtiax** [google/automl#796]
This isn't so much a bug report as a general request for advice.

Here's my problem set-up.  I want to detect humans (and no other classes) from overhead images.  I have a dataset of these images with labels, and a second dataset with only a few labels.  The two datasets are from slightly different conditions, but they are similar enough that I think training on the first should transfer at least somewhat to the second.  However, it seems like no matter what I try, any fine-tuning on the first dataset degrades performance on the second (compared to just using an off-the-shelf D0 or whatever).

Here's my current pipeline:

1) Starting from an Imagenet-trained efficientnet backbone, train a D0 model on a subset of COCO using only the person class.  In particular, set num_classes to be 2 in my hyperparameters.  The motivation for doing this is that it will give me a model whose box & class output layer is shaped correctly for the fine-tuning task.  The alternative is to just use the 80-class pretrained weights, but this requires re-initializing a fresh output layer, which seems undesirable.  (I've tried this way as well, but with no luck)

This model is trained for 50 epochs with an epoch size that corresponds to one pass through all images.  It performs well on a split of the COCO val set with only person annotations, so I believe that this section of the pipeline is working just fine.

2) Fine-tune on my overhead dataset.  To do this, I set up a .yaml hparams file that looks like this:

```
use_keras_model: False
num_classes: 2
label_map: {1: person}
autoaugment_policy: v1
var_freeze_expr: '(efficientnet|fpn_cells|resample_p6)'
learning_rate: 0.001
lr_warmup_init: 0.0001
```

I avoid using the keras model, as that was giving me trouble before (although it sounds like it may now be fixed).  I set num_classes to 2 (person + "background").  I use a autoaugment because my overhead dataset has less variety than I'd like.  I freeze the backbone, the bifpn, and the resampling layer (so all that's unfrozen is the class and box output layers).  I set a low learning rate and an even lower warmup init learning rate.

I then re-run the training command, but pointing to my new hparams yaml file, the fine-tuning training data, and I set the number of epochs to be 100.  The idea is that I've already done 50 in the original training and I want to do 50 more.  I'm not positive whether this is the correct approach - I don't have to do this sort of +50 trick when fine-tuning from one of the checkpoints provided in the repo.  Is there an intermediate step here that I should be doing to package up my initial model before fine-tuning?  I'm a little concerned that this may be the source of an error.

3) As fine-tuning progresses, eval on my second dataset (the one I want to transfer to).  What I observe is that even as performance on the first fine-tuning set goes up, performance on this second eval set goes down.  I would have expected a sort of U-shaped eval loss curve, where initial fine-tuning helps on the transfer set, but eventually we overfit and lose progress on the transfer set.  This doesn't seem to happen - even a few gradient descent steps on the fine-tuning set hurts performance on the transfer set.

If anyone can spot what I might be doing wrong, I'd be very apperciative.  Or if you've been able to fine-tune in a similar setting, and can post your working process, that'd also be super helpful.

**Comment by NikZak** [comment]
If you split the first dataset into train/dev - is eval working well on the dev set?

**Comment by nemtiax** [comment]
I tried the above experiment, and it seems that fine-tuning is also harmful on the dev set.  Some more details:

My dataset is HABBOF (http://vip.bu.edu/projects/vsns/cossy/datasets/habbof/), a set of overhead image with annotated people.  The original annotations are rotated boxes, so I've converted them to axis-aligned and put them in COCO-format.  The HABBOF set contains four splits, each of which is a contiguous video sequence.  Two splits take place in their "lab" room, and two in their "meeting" room.  I made the lab rooms my fine-tuning training set, and the meeting rooms my fine-tuning dev set.  (I could also intermix them, but this would result in near-identical images being in train and dev, because the sequences are contiguous videos, adjacent frames are very similar).

<img width="368" alt="Screen Shot 2020-09-23 at 12 40 40 PM" src="https://user-images.githubusercontent.com/6548233/94043215-92ecba00-fd9a-11ea-8255-e3a317bfed36.png">

Here is the AP on the dev split over a few epochs of fine-tuning on the train split.  This is qualitatively the same trajectory I get when evaluating on my second, separate dataset.

I've been wracking by brain for possible causes, here's what I can come up with:

-Continuing from epoch 50 could be causing some weirdness with the LR schedule?  Maybe I need to somehow package up my pretrained model so that it can be fine-tuned starting from epoch 0.

-Maybe there is some issue with batch norm layers doing something funky when frozen?  I added some print statements to the freeze_vars() method of efficientdet_arch.py, and everything looks fine.

-Maybe I'm incorrectly converting my data to tfrecords somehow.  My command is this:

```
PYTHONPATH=".:$PYTHONPATH"  python dataset/create_coco_tfrecord.py \
      --image_dir=habbof_lab/ \
      --object_annotations_file=habbof_annotations_lab.json \
      --output_file_prefix=tfrecord_habbof_lab/ \
      --num_shards=32
```
(and similarly for the meeting split).  Note that I do add a call to random.shuffle() on the images array in create_coco_tfrecord.py, because otherwise batches seem to consist of near-identical frames.  This may be breaking something? (Although if it were, I would expect it to also be breaking training set performance)

-Maybe I am incorrectly leaving some key hyperparameter on its default value.  I believe I'm matching the set-up for PASCAL fine-tuning in the README, but perhaps there's some other knob I need to turn for my person-only setting.  (see first post for my hparams settings)

Thanks for reading!

**Comment by nemtiax** [comment]
One more datapoint:

<img width="371" alt="Screen Shot 2020-09-23 at 1 20 07 PM" src="https://user-images.githubusercontent.com/6548233/94047053-8c147600-fd9f-11ea-8c71-3ce7ea971e0c.png">

The orange line shows the loss on the fine-tuning training set, and the blue line shows the loss on the fine-tuning dev set.  Note that the orange line is being computed on different batches of size 16 and so is spiky, whereas the blue line is being computed on a fixed set of 1024 validation images, and so is smooth.  What I find odd is that relatively minimal progress on the training loss (due to low learning rate, presumably) results in a consistent and large increase in the loss on the dev set.  If the problem were as simple as overfitting, I would have expected either much greater improvement in the training loss (due to successful memorization of training samples) or much lower degradation of the dev loss (due to relatively little change in the network performance).  I'm not sure how to explain the above trend, other than perhaps a bug somewhere either in the code base or (maybe more likely) in my train/eval pipeline?

**Comment by nemtiax** [comment]
Same experiment but without my additional shuffle() call when creating tfrecords:

<img width="360" alt="Screen Shot 2020-09-23 at 2 50 03 PM" src="https://user-images.githubusercontent.com/6548233/94056317-25498980-fdac-11ea-851d-ca2e13e4d809.png">
<img width="366" alt="Screen Shot 2020-09-23 at 2 50 09 PM" src="https://user-images.githubusercontent.com/6548233/94056319-25498980-fdac-11ea-8e1e-856bec296728.png">

Degradation on the dev set is a bit slower, but I suspect this is just because it's having more trouble learning the unshuffled training set.

**Comment by nemtiax** [comment]
<img width="363" alt="Screen Shot 2020-09-23 at 4 22 31 PM" src="https://user-images.githubusercontent.com/6548233/94065777-8deb3300-fdb9-11ea-9eed-70463de9de9e.png">

Same experiment, but now both the orange and blue line are my train split (I just set my validation_file_pattern argument to main.py to also point to the train split).  Seems very odd that eval reports rising loss on the training set while training reports falling loss.  Maybe an issue with saving or restoring the model for evaluation? (train_and_eval mode, which is what I'm running, seems to reload the model from the checkpoint every time it wants to eval)

**Comment by fitoule** [comment]
Try not freeze  '(efficientnet|fpn_cells|resample_p6)'
remove all to train all the layer
and remove autoaugment : it's worse 

When you have a pretty good result, then try autoaugment
and try to freeze 'efficientnet'

**Comment by nemtiax** [comment]
> Try not freeze '(efficientnet|fpn_cells|resample_p6)'
> remove all to train all the layer
> and remove autoaugment : it's worse
> 
> When you have a pretty good result, then try autoaugment
> and try to freeze 'efficientnet'

Thanks for the suggestions!  I'll give this a try and report back.  I was expecting that autoaugment might help with my specific dataset because of how uniform it is, but maybe it's doing more harm than good.  I was freezing all the earlier layers, both because that's what's recommended in the PASCAL example in the readme, and because I figure my dataset probably benefits from the extra regularization provided by freezing most of the network.  But maybe that's too aggressive, and I'd be better to only freeze the earliest parts.

Still, I would have expected that if these things are the problem, they'd only result in less-than-optimal fine-tuning, rather than breaking it entirely.  Although maybe an autoaugment issue would explain why my fine-tuning is giving worse performance on eval (since autoaugment would be off during eval).

**Comment by nemtiax** [comment]
Some initial results on the above experiment.

With unfreezing all layers, and no autoaugment, I can at least transfer across HABBOF splits.  In this experiment, I use HABBOF lab as my fine-tuning set (orange), and HABBOF meeting as my eval set (blue):

<img width="361" alt="Screen Shot 2020-09-24 at 12 16 54 PM" src="https://user-images.githubusercontent.com/6548233/94171988-eaa12900-fe5f-11ea-84be-5f5c1e4bf28c.png">
<img width="361" alt="Screen Shot 2020-09-24 at 12 17 02 PM" src="https://user-images.githubusercontent.com/6548233/94171989-eb39bf80-fe5f-11ea-9af0-dad9c4d6c2fe.png">

The AP still strikes me as quite low (HABBOF is a very easy dataset), but it's clearly still improving and might get there with enough epochs.  Given published results on HABBOF, I would hope to be able to get >90 AP when training on one split and testing on another.  Maybe with a bigger model and more training time...

I get significantly improved training loss compared to previous experiments, which is probably due to both a lack of augmentation (less training variety makes it easier), and unfreezing everything (more free parameters to fit the training set with).  The eval loss is still quite a bit above the training loss, but at least it's trending in the right direction now.

Two hiccups I've been running into, which are probably not significant problems, but I'm wondering if anyone has a solution:

1) Because I'm training my initial COCO network for 50 epochs, and then doing fine-tuning by resuming training with a new dataset and specifying 100 epochs, my learning rate schedule is starting from epoch 50.  I'd like to reset the epoch count and/or global step count so that fine-tuning is treated as starting from epoch 0.  Is there an easy way to edit my checkpoint to do this?

2) For some reason, when I run my fine-tuning in train_and_eval mode, it seems to enter a loop of evalling over and over with no training.  However, if I terminate that run, and then restart it (same exact command), it evals once and then starts alternating training and evalling as normal.  Does anyone know what causes this? Is it possible I'm not using train_and_eval mode correctly?

Anyway, I'm going to see if unfreezing everything and not using autoaugment allows me to transfer between my two datasets, and will report back.

Thanks for the help so far!

**Comment by nemtiax** [comment]
<img width="385" alt="Screen Shot 2020-09-24 at 1 37 05 PM" src="https://user-images.githubusercontent.com/6548233/94179813-107ffb00-fe6b-11ea-8309-a835c1c0cf7a.png">

Here are the results of trying to transfer from HABBOF to my actual test set.  The general shape of curve is not particularly surprising - it's the classic U-shape of fine-tuning where you see initial transfer success followed by overfitting the fine-tuning set.  What is surprising to me is how low the peak is.  I'm quite certain that fine-tuning (correctly) on HABBOF should transfer well to my test set, despite their differences.  I've tried another published model which was trained in this way (https://arxiv.org/abs/2005.11623) and it works excellently on my test set.  So I strongly suspect that the above lackluster performance is far below what efficientdet should be capable of on my data.  But I'm really out of ideas on what else to try.

**Comment by nemtiax** [comment]
Another thought - should I be specifying an image size for my evaluation somehow?  During training, presumably images are being resized to the D0 size (512x512?), but is this also happening during evaluation?  My raw HABBOF images are 2048x2048, so if they're being evaluated at that size, it might be a significant gap between training and testing.  Similarly, my test files are 1600x1200, so there's a similarly large difference in resolution, as well as a difference in aspect ratio.   The scale of objects in HABBOF is relatively uniform (it has a fixed camera position) so it certainly seems like the network would learn to very strongly prefer a particular object size.

Does evaluation mode automatically resize to the model's default or does it keep the raw input size?  If the latter, is there a way to tell it to resize (I guess I could just manually resize my images before making tfrecords)?

Also, should I be overriding the default anchor sizes in this sort of situation?  I didn't want to mess with too many knobs at once, but it seems like it might be better to not have the full range of anchor sizes if no objects are going to use them.

**Comment by nemtiax** [comment]
One more data point.  Fine-tuning (with nothing frozen) for a sufficient amount of epochs on HABBOF lab and then evaluating on HABBOF lab gives me >0.99 AP50.  This makes me think it's very unlikely that there is any issue with different image sizes or whatever in the eval pipeline.

However, evaluating this same model on HABBOF meeting (i.e. the same dataset just a different room) gives merely 0.66 AP50. This indicates to me that I'm getting severe overfitting.  This is in line with what I would expect from a uniform dataset like HABBOF and no weight freezing.  So I don't think that just unfreezing all the weights is likely to yield a solution to my problem.

**Comment by SiBensberg** [comment]
>     1. Because I'm training my initial COCO network for 50 epochs, and then doing fine-tuning by resuming training with a new dataset and specifying 100 epochs, my learning rate schedule is starting from epoch 50.  I'd like to reset the epoch count and/or global step count so that fine-tuning is treated as starting from epoch 0.  Is there an easy way to edit my checkpoint to do this?

Give your model a new name when training on the second dataset. This will reset the initial epoch count. Obviously you need to assign your "first stage model" as starting checkpoint: ``` --ckpt=efficientdet-d1-first-stage ```

>      So I don't think that just unfreezing all the weights is likely to yield a solution to my problem.

With other architectures I made similiar experiences. I mostly freeze at least the pre-trained backbone. The performance loss was mostly low at about 0.02. I tackled it with some low lr epochs at the end with everything unfreezed.

**Comment by nemtiax** [comment]
> Give your model a new `--model_dir`when training on the second dataset. This will reset the initial epoch count. Obviously you need to assign your "first stage model" as starting checkpoint: `--ckpt=your-first-model-dir`

This doesn't work for me - I always specify a new model_dir for the fine-tuning, but it does not reset the epoch count.  Is there anything else I need to change?

**Comment by SiBensberg** [comment]
you are right. In the checkpoints is a tensor named 'global_step'. How to reset it to 0? 

I read it with the following code snippet and tf 2.2:

```
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import tensorflow as tf

from tensorflow.python.training import py_checkpoint_reader


file_name = 'model.ckpt-300000'
tensor_name = 'global_step'

reader = py_checkpoint_reader.NewCheckpointReader(file_name)

if not reader.has_tensor(tensor_name):
    print("Tensor %s not found in checkpoint" % tensor_name)

var_to_shape_map = reader.get_variable_to_shape_map()
var_to_dtype_map = reader.get_variable_to_dtype_map()

print("tensor: %s (%s) %s" %
      (tensor_name, var_to_dtype_map[tensor_name].name,
       var_to_shape_map[tensor_name]))
print(reader.get_tensor(tensor_name))
```

**Comment by matankley** [comment]
Hey, have you found a solution for resetting the global step ?

**Comment by kevinkvothe** [comment]
I can confirm that extracting a model using `model_inspect.py --runmode=saved_model_infer \` and using the variables.data and variables.index files to substitute the model.data and model.index files on the checkpoint resets the global step to 0.

**Comment by SiBensberg** [comment]
@kevinkvothe Thank you for that solution! Will try out asap

**Comment by mingxingtan** [comment]
My bad, the global_step should be excluded from pretrained checkpoints.
I have fixed it in https://github.com/google/automl/commit/9c36d2cf66feee328d98dca02700ba708ba1d643
